snuba/query/parser/conditions.py

```diff
@@ -178,9 +179,12 @@ def unpack_array_condition_builder(
         )
 
     def simple_condition_builder(lhs: Expression, op: str, literal: Any) -> Expression:
-        return binary_condition(
-            None, OPERATOR_TO_FUNCTION[op], lhs, preprocess_literal(op, literal)
-        )
+        if literal is not None:
+            return binary_condition(
+                None, OPERATOR_TO_FUNCTION[op], lhs, preprocess_literal(op, literal)
+            )
+        else:
+            return unary_condition(None, OPERATOR_TO_FUNCTION[op], lhs)
```

Comment:
Are you sure we can assume that we have a unary condition here? Isn't "col = None" a valid condition that shouldn't be parsed as unary

Reply:
I don't think that's a valid condition. It would be `["col", "=", "NULL"]`.

Reply:
Also if you give snuba`["level", "=", null]` which turns into `["level", "=", None]` it becomes (without ast) for clickhouse `SELECT event_id FROM sentry_local PREWHERE project_id IN (1) WHERE level =  AND timestamp >= toDateTime('2020-06-04T17:45:26', 'Universal') AND timestamp < toDateTime('2020-06-09T17:45:26', 'Universal') AND deleted = 0 LIMIT 0, 1000`

`level =  AND`. instead of `level = NULL`.
So I think both the legacy query parsing and legacy language are not correct with regards to conditions that include null as literal.

Reply:
I can add an assertion that prevents binary conditions to be provided to unary operators and vice versa though

Reply:
I'm quite sure the first example `["col", "=", "NULL"]` is supposed to check for equality with the string "NULL", not converted into IS NULL.

Reply:
Sorry rereading my comment I think I was not clear. That still works anyway. The condition this method would receive would be `"col", "=", "NULL"` not  `"col", "=", None` so it would be turned into a binary condition.
```
SELECT event_id FROM sentry_local PREWHERE project_id IN (1) WHERE level = 'NULL' AND timestamp >= toDateTime('2020-06-04T18:15:42', 'Universal') AND timestamp < toDateTime('2020-06-09T18:15:42', 'Universal') AND deleted = 0 LIMIT 0, 1000
```

Reply:
Assertion added so we always parse only valid conditions.

Reply:
Could we enforce this in the schema instead?

Reply:
I think it should be possible. The only issue with that is that, if we start by enforcing it here, we would see warnings if someone is sending queries to the old query processing pipeline that somehow work, because the ast parsing is still not blocking, so we would know whether it is safe to do so. If I change the schema instead those queries would fail right away. What do you think ?

Reply:
If you think that's a risk it's probably worth doing the schema change as a follow up then so we don't break any accidentally working code. If the assertion fails it gets sent to Sentry right? Did you already try running the Sentry tests against this change? I'd hope that most of these cases would be covered there.

Reply:
They are sent to sentry (as warnings). I did not try to run the sentry tests. Will try now.

Reply:
No failure due to this. I would be happy to make the change to the schema as an immediate followup as soon as we confirm there is no warning in the dashboard.